### DISCUSSION TITLE
Configure renovate within a monorepo

### DISCUSSION THREAD

**Discussion by achimgrimm** [renovatebot/renovate#15711] [Request Help]
Hi,
I have a monorepo that needs different renovate configs per module.
Until now I had two ideas to approach this:

1) Use package rules to and `matchPaths` to select the modules.
This has the downside, that I can not use all my presets as I would be able to in a single repo.
see https://github.com/renovatebot/renovate/issues/7373

2) Use multiple renovate configs - one per module.
As I understand the documentation this is not a valid use case either.
see https://docs.renovatebot.com/configuration-options/

How could I approach such a problem?
Happy to get a helping hand here.

**Comment by rarkins** [comment]
You should go with option (1) because as you note, option (2) is not valid

**Reply by achimgrimm** [reply]
Is there any chance to use regular presets for option (1)? I understand the linked issue, that this is not possible.
I would need to include the package rule directly and am not able to reuse the presets.

I just want to be sure, that I understand this correctly and am not missing any optimisation. 
Thank you in advance.

**Reply by rarkins** [reply]
Can you give an example of what you mean?

**Reply by achimgrimm** [reply]
Yes, of course.

This is an example config in the mono repo.
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "my-local-repo/renovate-presets:base"
  ],
  "baseBranches": [
    "main"
  ],
  "packageRules": [
    {
      "groupName": "Frontend",
      "matchPaths": ["frontend/"]
    },
    {
      "groupName": "Backend",
      "matchPaths": ["backend/"],
      "extends": [
        "my-local-repo/renovate-presets:java-17",
        "my-local-repo/renovate-presets:postgresql-11"
      ]
    }
  ]
}
```

This is the java-17 preset - as an example:
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "description": "Pin to Java version 17",
  "packageRules": [
    {
      "matchPackagePrefixes": [
        "eclipse-temurin"
      ],
      "allowedVersions": "<18"
    }
  ]
}
```

This is the log message during the validation:
```
 INFO: Validating renovate.json
 INFO: Throwing preset error
       "validationError": undefined
 INFO: renovate.json is not valid Renovate config
       "err": {
         "message": "config-validation",
         "stack": "Error: config-validation\n    at resolveConfigPresets (/dist/.nvm/versions/node/v16.13.0/lib/node_modules/renovate/dist/config/presets/index.js:258:35)\n    at async validateConfig (/dist/.nvm/versions/node/v16.13.0/lib/node_modules/renovate/dist/config/validation.js:277:45)\n    at async validate (/dist/.nvm/versions/node/v16.13.0/lib/node_modules/renovate/dist/config-validator.js:26:17)\n    at async /dist/.nvm/versions/node/v16.13.0/lib/node_modules/renovate/dist/config-validator.js:50:17"
       }

```

**Reply by rarkins** [reply]
So your goal is that this config:

```
    {
      "groupName": "Backend",
      "matchPaths": ["backend/"],
      "extends": [
        "my-local-repo/renovate-presets:java-17"
      ]
    }
```

Becomes this config?

```
    {
      "groupName": "Backend",
      "matchPaths": ["backend/"]
    },
    {
      "matchPaths": ["backend/"],
      "matchPackagePrefixes": [
        "eclipse-temurin"
      ],
      "allowedVersions": "<18"
    }
```

It's not impossible to do, but it's pretty complicated logic. e.g. I've needed to merge the matchX options and remove the groupName.

It would be simpler with this instead:

```
    {
      "matchPaths": ["backend/"],
      "groupName": "Backend"
    },
    {
      "matchPaths": ["backend/"],
      "extends": [
        "my-local-repo/renovate-presets:java-17"
      ]
    }
```
..although still complex to resolve. But we do have some code like this we attempt to massage.

**Reply by achimgrimm** [reply]
I am not sure if I fully understand the difference of the two versions to be honest. 

The goal is to reuse the configurations for the presets to reduce code replication. So both should be fine. :)

**Reply by aldenquimby** [reply]
@rarkins I am hitting the same issue and don't understand your most recent response. I have a monorepo with ~10 separate packages that currently uses dependabot. I'm trying to move to renovate to control grouping and reduce PR noise, ideally with a config like this:
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["schedule:nonOfficeHours"],
  "labels": ["dependencies"],
  "assigneesFromCodeOwners": true,
  "packageRules": [
    {
      "matchPackageNames": ["node", "npm"],
      "enabled": false
    },
    {
      "matchPackageNames": ["@types/node"],
      "allowedVersions": "^14"
    },
    {
      "matchPaths": ["src/package1", "src/package2", "src/package3"],
      "extends": ["config:js-app"]
    },
    {
      "matchPaths": ["src/package2"],
      "matchPackageNames": ["axios"],
      "allowedVersions": "0.21.4"
    },
    {
      "matchPaths": ["src/package4", "src/package5", "src/package6"],
      "extends": ["config:js-lib", "npm:unpublishSafe"]
    }
  ],
  "hostRules": [
    {
      "matchHost": "https://npm.pkg.github.com/",
      "hostType": "npm",
      "encrypted": { "token": "TOKEN_HERE" }
    }
  ],
  "npmrc": "@mycorp:registry=https://npm.pkg.github.com/",
  "npmrcMerge": true
}
```

Is there any way to apply an "extends" within packageRules?

The only option I can see is to copy/paste the implementations of `:pinAllExceptPeerDependencies`, `: pinOnlyDevDependencies`, and any other presets we use inline each packageRule

**Reply by rarkins** [reply]
When you configure `config:js-lib`, why are you doing that? e.g. is it because you want to preserve version ranges/constraints? And similar for js-app it's because you want to pin versions? Because those can be controlled using the `rangeStrategy` field within `packageRules`
